File: 74 Search a 2D Matrix/search_2D_matrix.js
class Solution {
    searchInRow(matrix, target, row) {
        let n = matrix[0].length;
        let start = 0, end = n - 1;
        
        while (start <= end) {
            let mid = start + Math.floor((end - start) / 2);
            
            if (target === matrix[row][mid]) {
                return true;
            } else if (target > matrix[row][mid]) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        return false;
    }
    
    searchMatrix(matrix, target) {
        let m = matrix.length;
        let n = matrix[0].length;
        
        let startRow = 0, endRow = m - 1;
        
        while (startRow <= endRow) {
            let midRow = startRow + Math.floor((endRow - startRow) / 2);
            
            if (target >= matrix[midRow][0] && target <= matrix[midRow][n - 1]) {
                return this.searchInRow(matrix, target, midRow);
            } else if (target > matrix[midRow][n - 1]) {
                startRow = midRow + 1;
            } else {
                endRow = midRow - 1;
            }
        }
        return false;
    }
}

const solution = new Solution();

const matrix = [
    [1, 3, 5, 7],
    [10, 11, 16, 20],
    [23, 30, 34, 60]
];

console.log(solution.searchMatrix(matrix, 3));   
console.log(solution.searchMatrix(matrix, 13));  
console.log(solution.searchMatrix(matrix, 34));  

File: 74 Search a 2D Matrix/search_2D_matrix.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool searchInRow(vector<vector<int>> & matrix, int target, int row){
        int n = matrix[0].size();  // n = number of columns
        int start = 0, end = n - 1;

        while(start <= end){
            int mid = start + (end - start)/2;

            if(target == matrix[row][mid]){
                return true;
            }else if(target > matrix[row][mid]){
                start = mid + 1;
            }else{
                end = mid - 1;
            }
        }
        return false;
    }

    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();     // m = number of rows
        int n = matrix[0].size();  // n = number of columns

        int startRow = 0, endRow = m-1;
        while(startRow <= endRow){
            int midRow = startRow + (endRow - startRow)/2;

            if(target >= matrix[midRow][0] && target <= matrix[midRow][n-1]){
                return searchInRow(matrix, target, midRow);
            }else if(target >= matrix[midRow][n-1]){
                startRow = midRow + 1;
            }else{
                endRow = midRow - 1;
            }
        }
        return false;
    }
};
int main() {
    Solution solution;
    
    vector<vector<int>> matrix = {
        {1, 3, 5, 7},
        {10, 11, 16, 20},
        {23, 30, 34, 60}
    };

    int target01 = 3; 
    int target02 = 13;
    int target03 = 34;
    
    cout << solution.searchMatrix(matrix, target01) << endl;   
    cout << solution.searchMatrix(matrix, target02) << endl;   
    cout << solution.searchMatrix(matrix, target03) << endl;   
    
    return 0;
}